lib/engine/game/g_1812/game.rb

```diff
@@ -446,12 +448,13 @@ def revenue_for(route, stops)
           train = route.train
           corp = route.corporation
 
-          revenue += 10 if corp.assigned?(p4_company) && stops.find { |s| C18_HEX.include?(s.hex.id) }
+          revenue += 10 if corp.assigned?(p4_company) && stops.any? { |s| s.hex.id == C18_HEX }
 
-          revenue += 10 if g_train?(train) && corp.assigned?(p3_company) && stops.any? { |s| F3_PORT.include?(s.hex.id) }
-          revenue += 10 if g_train?(train) && corp.assigned?(p8_company) && stops.any? { |s| F3_PORT.include?(s.hex.id) }
-          revenue += 10 if g_train?(train) && corp.assigned?(p9_company) && stops.any? { |s| H9_PORT.include?(s.hex.id) }
-          revenue += 20 if g_train?(train) && corp.assigned?(p12_company) && stops.any? { |s| G6_PORT.include?(s.hex.id) }
+          { p3_company => F3_PORT, p8_company => F3_PORT, p9_company => H9_PORT, p12_company => G6_PORT }.each do |company, port|
+            revenue += (company == p12_company ? 20 : 10) if g_train?(train) && corp.assigned?(company) && stops.any? do |s|
+                                                               s.hex.id == port
+                                                             end
```

Comment:
Yeah, this isn't clearer. The `HexBonus` ability _almost_ matches what you need here (and does exactly for P4), you just need some way of testing that the train is a goods train. Maybe add a `trains` array to that ability, then in `revenue_for` test both that you have a matching hex and a matching train.

Reply:
Ok, I did some clean up and improved some methods I think. 

commit: `cleans up revenue_for and associated methods`